File: lib_base/src/main/java/com/android/base/widget/viewpager/PageNumberView.java
package com.android.base.widget.viewpager;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.util.AttributeSet;
import android.view.View;

import com.android.base.R;
import com.android.base.utils.android.UnitConverter;


public class PageNumberView extends View implements IPagerNumberView {

    private static final String DIVISION = "/";

    private int mPageSize;
    private int mPosition;

    private int mCenterX;

    private Paint mPaint;
    private int mTextHeight;
    private int mTextBaseLine;

    public PageNumberView(Context context) {
        this(context, null);
    }

    public PageNumberView(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public PageNumberView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);

        TypedArray typedArray = context.obtainStyledAttributes(attrs, R.styleable.PageNumberView);

        /*是否可以缩放*/
        int color = typedArray.getColor(R.styleable.PageNumberView_pnv_text_color, Color.BLACK);
        /*用于支持5.0的transition动画*/
        int size = typedArray.getDimensionPixelSize(R.styleable.PageNumberView_pnv_text_size, UnitConverter.spToPx(14));

        typedArray.recycle();

        mPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
        mPaint.setTextSize(size);
        mPaint.setColor(color);

        Paint.FontMetricsInt fontMetricsInt = mPaint.getFontMetricsInt();
        mTextHeight = fontMetricsInt.bottom - fontMetricsInt.top;
        mTextBaseLine = mTextHeight - fontMetricsInt.bottom;
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        String text = String.valueOf(mPageSize);
        float textWidth = mPaint.measureText(text);
        float divisionWidth = mPaint.measureText(DIVISION);
        setMeasuredDimension((int) (textWidth * 2 + divisionWidth), mTextHeight);
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        float divisionHalfSize = mPaint.measureText(DIVISION) / 2;
        float left = mCenterX - divisionHalfSize;
        float right = mCenterX + divisionHalfSize;
        String positionText = String.valueOf(mPosition);
        String sizeText = String.valueOf(mPageSize);
        canvas.drawText(DIVISION, left, mTextBaseLine, mPaint);
        canvas.drawText(positionText, left - mPaint.measureText(positionText), mTextBaseLine, mPaint);
        canvas.drawText(sizeText, right, mTextBaseLine, mPaint);
    }

    @Override
    protected void onSizeChanged(int w, int h, int oldw, int oldh) {
        super.onSizeChanged(w, h, oldw, oldh);
        mCenterX = w / 2;
    }

    @Override
    public void setPageScrolled(int position, @SuppressWarnings("UnusedParameters") float positionOffset) {
        mPosition = position;
        if (mPageSize > 1) {
            if (mPosition == 0) {
                mPosition = mPageSize;
            } else if (mPosition == mPageSize + 1) {
                mPosition = 1;
            }
        } else {
            mPosition = mPageSize;
        }
        invalidate();
    }

    @Override
    public void setViewPager(ZViewPager viewPager) {
        //no op
    }

    @Override
    public void setPageSize(int pageSize) {
        mPageSize = pageSize;
        mPosition = pageSize;//只有一個的時候
        requestLayout();
    }

}

File: lib_common_ui/src/main/java/me/ztiany/widget/viewpager/PageNumberView.kt
package me.ztiany.widget.viewpager

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.util.AttributeSet
import android.view.View
import com.android.sdk.ui.R
import me.ztiany.widget.common.dip


class PageNumberView @JvmOverloads constructor(
        context: Context,
        attrs: AttributeSet? = null,
        defStyleAttr: Int = 0
) : View(context, attrs, defStyleAttr), IPagerNumberView {

    private var mPageSize: Int = 0
    private var mPosition: Int = 0

    private var mCenterX: Int = 0

    private val mPaint: Paint
    private val mTextHeight: Int
    private val mTextBaseLine: Int

    init {

        val typedArray = context.obtainStyledAttributes(attrs, R.styleable.PageNumberView)

        /*是否可以缩放*/
        val color = typedArray.getColor(R.styleable.PageNumberView_pnv_text_color, Color.BLACK)
        /*用于支持5.0的transition动画*/
        val size = typedArray.getDimensionPixelSize(R.styleable.PageNumberView_pnv_text_size, dip(14))

        typedArray.recycle()

        mPaint = Paint(Paint.ANTI_ALIAS_FLAG)
        mPaint.textSize = size.toFloat()
        mPaint.color = color

        val fontMetricsInt = mPaint.fontMetricsInt
        mTextHeight = fontMetricsInt.bottom - fontMetricsInt.top
        mTextBaseLine = mTextHeight - fontMetricsInt.bottom
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        val text = mPageSize.toString()
        val textWidth = mPaint.measureText(text)
        val divisionWidth = mPaint.measureText(DIVISION)
        setMeasuredDimension((textWidth * 2 + divisionWidth).toInt(), mTextHeight)
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        val divisionHalfSize = mPaint.measureText(DIVISION) / 2
        val left = mCenterX - divisionHalfSize
        val right = mCenterX + divisionHalfSize
        val positionText = mPosition.toString()
        val sizeText = mPageSize.toString()
        canvas.drawText(DIVISION, left, mTextBaseLine.toFloat(), mPaint)
        canvas.drawText(positionText, left - mPaint.measureText(positionText), mTextBaseLine.toFloat(), mPaint)
        canvas.drawText(sizeText, right, mTextBaseLine.toFloat(), mPaint)
    }

    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        super.onSizeChanged(w, h, oldw, oldh)
        mCenterX = w / 2
    }

    override fun setPageScrolled(position: Int, positionOffset: Float) {
        mPosition = position
        if (mPageSize > 1) {
            if (mPosition == 0) {
                mPosition = mPageSize
            } else if (mPosition == mPageSize + 1) {
                mPosition = 1
            }
        } else {
            mPosition = mPageSize
        }
        invalidate()
    }

    override fun setViewPager(viewPager: BannerViewPager) {
        //no op
    }

    override fun setPageSize(pageSize: Int) {
        mPageSize = pageSize
        mPosition = pageSize//只有一個的時候
        requestLayout()
    }

    companion object {
        private const val DIVISION = "/"
    }

}